### DISCUSSION TITLE
IPsec packet offload IKE by-pass policies

### DISCUSSION THREAD

**Discussion by emeel97** [strongswan/strongswan#1597] [Q&A]
Hi,

After closing https://github.com/strongswan/strongswan/pull/1462 and merging the code to the master branch we noticed an issue.
Before the merge we got asked to test this feature with our hardware and we did that using a POC code changes to the kernel, at a time it worked fine (since we had a lot of hardcode).
After implementing the relevant parts in the kernel and the mlx5 driver properly we faced an issue in the process of installing the IKE bypass policies.

Debugging the issue showed that in the process of adding port bypass (add_port_bypass routine in strongswan code)
we try to add an IKE by-pass policies for both directions (in & out) , and if we fail to add one of the policies we stop the whole process and we fail.

In our case we failed to add the dir in (POLICY_IN) bypass policy (since the mlx5 driver blocks such flow and only allows the dir out bypass policies), we tried not adding the POLICY_IN bypass policy which fixed the issue (we managed to create the dir out policy and the IKE daemon managed to exchange keys and the session got created with packet offload enabled).

The above led to asking the following question:
- I'm wondering why do we need the dir in (POLICY_IN) bypass policy if we are managing to work with out it?
- is there any added value for adding a dir in (POLICY_IN)  bypass policy that we might be missing? 
- since having only a dir out (POLICY_OUT) bypass policy works fine shouldn't we adapt the strongswan code to not fail completely in case it fails to add the dir in (POLICY_IN) bypass policy?

what are your thoughts?

Thanks!

**Comment by tobiasbrunner** [comment]
> * I'm wondering why do we need the dir in (POLICY_IN) bypass policy if we are managing to work with out it?
> 
> * is there any added value for adding a dir in (POLICY_IN)  bypass policy that we might be missing?

You obviously need both. How would inbound packets be allowed after actual IPsec policies are installed? The point of the latter is to enforce that inbound packets that match them were encrypted. But if the IKE packets match them (e.g. if the remote traffic selector is `0.0.0.0/0` or even just matching the peer's IP address), any unprotected IKE packets would get dropped. So you need an inbound policy that allows such unencrypted packets to port 500/4500.


> * since having only a dir out (POLICY_OUT) bypass policy works fine

Did you try initiating a rekeying from the other peer after establishing an SA with traffic selectors that overlap with the peer's IP? Or try to initiate a completely new SA from a different peer if the traffic selector overlaps with the IP of that peer?

Or did you try offloading a drop policy for `0.0.0.0/0` first and then initiate an SA from a peer?

**Reply by emeel97** [reply]
I though during re-key the ipsec would be removed so then only POLICY_OUT would be enough.
anyway I will allow our driver to add dir in (POLICY_IN) bypass policy .

Thanks tobias

**Reply by tobiasbrunner** [reply]
IKE rekeying (with IKEv2) is independent of IPsec SAs and policies, and there are other IKE messages (e.g. DPDs) that also have to bypass the installed polices.

CHILD_SA rekeying happens overlapping to prevent traffic interruptions. And the IPsec policies stay installed all the time (in particular the inbound policy is not changed at all as we don't set an SPI on it; the outbound policy is updated with the SPI of the new outbound SA).

When the replacement CHILD_SA is negotiated, the new inbound SA is installed immediately. The installation of the outbound SA can be delayed (e.g. on responders whose IPsec stack does not allow referencing specific SAs via outbound policy). On the responder, the old outbound SA is also not immediately deleted when responding to the CREATE_CHILD_SA request, but only when the delete for the old CHILD_SA is received (at that point we know the initiator has received the CREATE_CHILD_SA response and has installed the inbound SA). This also updates the SPI in the OUT policy so it references the new SA. Both initiator and responder of the rekeying keep the inbound SA installed a bit longer so delayed ESP packets can still be processed.
